crates/subspace-service/src/lib.rs

```diff
@@ -503,6 +522,12 @@ where
         }
     };
 
+    task_manager.spawn_essential_handle().spawn_essential(
+        "bundle-collector",
+        None,
+        Box::pin(bundle_collector.run()),
+    );
```

Comment:
> It's hard to read what this task is for because it's invisible in the architecture and the name is too vague.

The naming could definitely be improved, but I am not sure we should put some of the worker logic here since it is too verbose compared to other workers (like `root-block-archiver`).

Reply:
Well, it's merely about 10 lines, I don't think that's too much, the initialize and update are two function calls. But this change will make the reader easily see what's happening here without looking into the actual worker.

Reply:
> But this change will make the reader easily see what's happening here without looking into the actual worker.

I am not sure this is what we need to achieve here as this function already ~300 locs, and these codes should be the implemented detail that should keep inside its won crate and it is not always interested by people who come to this function.

Indeed, it may be nice for some newcomers to the codebase to have a look at this function and know what component/worker a node consists of and what they roughly doing. But besides that, perhaps most of the time people who come to this function just want to scroll the mouse away from these details and find the exact component/worker they are interested (within ~300 locs).

But I am not strongly opposed though and will make these changes if you insist.

Reply:
It's not a hard requirement, but why not if it can be achieved easily?

No strong opinion either as long as you can make it descriptive and understandable, I can guess what the other tasks are for without looking into them, but not this one.

Reply:
The bundle-collector worker logic is outlined in `new_full`.

> the initialization of stored bundles in the last k is implicit on the first imported block notification, which is not intuitive and incorrect even because new bundles can be received before the very first block notification

But this part is unchanged because I found that as the node is just started, it will need a major sync to sync block from other peers, and before that finish, it can only collect bundle from some old block instead the last K block, so it doesn't help much to initialize the bundles on startup.

Also, I found that the `import_notification_stream` may start to notify block import before it actually synced to the tip of the chain (more specifically, 5 blocks before the tip of the chain):

https://github.com/paritytech/substrate/blob/2a636e26e9c200bb77910e4956210bce2bef0e22/client/network/sync/src/lib.rs#L489-L502

and as the transaction in txn pool will be re-validated after a new block is imported, so it may be okay to initialize on the first imported block notification.

Reply:
>as the transaction in txn pool will be re-validated after a new block is imported, so it may be okay to initialize on the first imported block notification.

Well, the problem I mentioned before still persists, it's possible to include the duplicated bundle in the very first block if it's not initialized. So we should initialize it only when the node is not in major sync on startup.

Reply:
The node will not building block while in major sync as the slot worker will skip propose slot. After the major sync finish, the node still have 5 more blocks waiting to import, while these blocks import the bundle will be initialized and the txn will be re-validated.

But indeed there is a corner case where the major sync just finished and before next block import notification come, the next slot come first and the node just win the challange and build the block without duplicated-bundle check, then created a fork that at least 5 block behide the best fork.

And I think `initialize it only when the node is not in major sync on startup` doesn't do any help to this case.

Reply:
> After the major sync finish, the node still have 5 more blocks waiting to import

The node may or may not have 5 more blocks to import, say, you just restart the node when it's already synced to the tip, in which case it has no newer blocks to import before producing the next block.

What we're trying to achieve is essentially to perform the duplicate bundle check to prevent them into the newer blocks, which requires whenever the major sync is false (slot worker will attempt to claim the slot) the bundles in last K should be available.